### PR TITLE
Move compressed marshaling away from the stdlib and into the compilers (trunk version)

### PR DESCRIPTION
**Pull request by xavierleroy**
This is the "forward port" of #12734 from the 5.1 branch to the trunk.

Besides adapting to recent changes in extern.c and in the top Makefile, this PR also proposes a different way to auto-link compressed marshaling support when ocamlcommon.cmxa is used:
- a new C library libcomprmarsh.a is built (from runtime/zstd.n.o) and installed along with libasmrun.a
- ocamlcommon.cmxa includes instructions to link with -lcomprmarsh and then with -lzstd.

@dra27 and I did not want to follow this approach in a minor release like 5.1.1, as it introduces a new installed file, which could upset packagers.  But it seems appropriate for a major release like 5.2.

This is the only significant change from #12734.  Compare 833ca02315 (this PR) with 9398511da3 (the 5.1.1 solution).

### CODE REVIEW

**Comment by xavierleroy** [comment]
> a new C library libcomprmarsh.a is built (from runtime/zstd.n.o) and installed along with libasmrun.a

Late thought: libcomprmarsh.a could probably just as well reside in compilerlibs/ , both during build and after installation.  Keep this possibility in mind during review.

**Comment by nojb** [comment]
> this PR also proposes a different way to auto-link compressed marshaling support

I understand this choice corresponds to the second bullet point in https://github.com/ocaml/ocaml/pull/12734#discussion_r1390452993. Was this chosen because of simplicity, to avoid changing the way the compiler works? And regardless of this PR, do we agree that the change mentioned in the first bullet point would be an improvement in general?

**Comment by xavierleroy** [comment]
> I understand this choice corresponds to the second bullet point in https://github.com/ocaml/ocaml/pull/12734#discussion_r1390452993. 

Yes!  You're right I should have been more explicit.

> Was this chosen because of simplicity, to avoid changing the way the compiler works? And regardless of this PR, do we agree that the change mentioned in the first bullet point would be an improvement in general?

For reference: the first bullet point is about changing `ocamlopt -a` so that `.o` arguments are added to the generated `.a` code archive instead of being recorded as "to be linked later" in the generated `.cmxa` archive descriptor.

I'm not 100% comfortable with this change because there's no way to change `ocamlc -a` in the same manner: it does not create a `.a` code archive, so there's no place to store the `.o` arguments...  So, we would be adding a feature that is usable only for natively-compiled libraries.  I really think all OCaml libraries should be compiled both to native-code and to bytecode, and remain usable in both contexts.

At any rate, the "first bullet" approach that is followed in this PR (an explicit libcomprmarsh.a library, autolinked from ocamlcommon.cmxa) is a step towards an otherlibs/compression library providing compressed marshaling and other data compression facilities to the users.  (Stay tuned!)

**Comment by gasche** [comment]
@hhugo is there a possibility that you would be interested in reviewing this PR?

File: runtime/zstd.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    rc = ZSTD_compressStream2(ctx, &out, &in,
+                              input == NULL ? ZSTD_e_end : ZSTD_e_continue);
+  } while (! (input == NULL && rc == 0));
+  output->end = output->data + out.pos;
+  *extern_output_first = output_head;
+  ZSTD_freeCCtx(ctx);
+  return true;
+oom2:
+  /* The old output blocks that remain to be freed */
```
Comment (damiendoligez) [inline runtime/zstd.c]:
```suggestion
  /* The old input blocks that remain to be freed */
```

**Review by damiendoligez — APPROVED** [review]
Looks good to me, I found only a typo.

**Comment by damiendoligez** [comment]
> Late thought: libcomprmarsh.a could probably just as well reside in compilerlibs/ , both during build and after installation. Keep this possibility in mind during review.

IIUC, this would make sense with the current version, but you'd have to move it back to provide compression for general marshalling, right?

File: runtime/zstd.c (hunk 1 above)
Comment (xavierleroy) [reply runtime/zstd.c]: Technically, `caml_zstd_compress` transforms uncompressed output blocks (from the marshaler) in compressed output blocks, so both its input and output are "output blocks" :-)  I'll rephrase the comments in terms of "uncompressed blocks" and "compressed blocks".

**Comment by xavierleroy** [comment]
Thanks for the feedback!  I merged without further delay because the bootstrap in this PR is a pain to re-do properly.  (It is absolutely necessary to disable ZSTD compression until the new compilers are stable.)  I will now attempt a backport to 5.2.

**Comment by xavierleroy** [comment]
Cherry-picked to 5.2.  The bootstrap had to be redone for 5.2, so I cherry-picked each commit individually.  We end up with cherry-picks of cherry-picks... I hope everything is OK.  That's what happens when urgent PRs are not reviewed before a release branch is cut.

**Pull request merged**